pygfx/objects/_ruler.py

```diff
@@ -0,0 +1,489 @@
+        return self._points
+
+    # Note: text should also be here eventually
+
+    # -- Main properties
+
+    @property
+    def start_pos(self):
+        """The start posision of the ruler as a 3-tuple or 3-element array.
+
+        Note that the ruler's transform also affects positioning, but should
+        generally not be used.
+        """
+        return self._start_pos
+
+    @start_pos.setter
+    def start_pos(self, pos):
+        start_pos = np.array(pos, np.float64).reshape((3,))
+        if start_pos.shape != (3,):
+            raise ValueError("Ruler.start_pos must be a 3-element position.")
+        self._start_pos = start_pos
+        self._end_value = None
+
+    @property
+    def end_pos(self):
+        """The end posision of the ruler as a 3-tuple or 3-element array."""
+        return self._end_pos
+
+    @end_pos.setter
+    def end_pos(self, pos):
+        end_pos = np.array(pos, np.float64).reshape((3,))
+        if end_pos.shape != (3,):
+            raise ValueError("Ruler.end_pos must be a 3-element position.")
+        self._end_pos = end_pos
+        self._end_value = None
+
+    @property
+    def start_value(self):
+        """The value of the ruler at the start position (i.e. the offset)."""
+        return self._start_value
+
+    @start_value.setter
+    def start_value(self, value):
+        self._start_value = float(value)
+        self._end_value = None
+
+    @property
+    def end_value(self):
+        """The value at the end of the ruler (read-only)."""
+        # Little caching mechanic. Props that affect the end_value set ._end_value to None
+        if self._end_value is None:
+            self._end_value = (
+                np.linalg.norm(self._end_pos - self._start_pos) + self._start_value
+            )
+        return self._end_value
+
+    @property
+    def ticks(self):
+        """The ticks to show.
+
+        Can be:
+
+        * ``None`` for automatic ticks.
+        * ``dict`` for explicit ticks (float -> str).
+        * ``list`` / ``tupple`` / ``ndarray`` for a list of values.
+
+        """
+        return self._ticks
+
+    @ticks.setter
+    def ticks(self, ticks):
+        if ticks is None:
+            self._ticks = None
+        elif isinstance(ticks, dict):
+            # Copy the object, resolving keys and values to float and str
+            self._ticks = {float(k): str(v) for k, v in ticks.items()}
+        elif isinstance(ticks, (tuple, list, np.ndarray)):
+            self._ticks = [float(x) for x in ticks]
+        else:
+            raise TypeError("Ruler.ticks must be None, dict or list(like).")
+
+    # -- Properties for tweaking
+
+    @property
+    def tick_format(self):
+        """The format to display the tick values."""
+        return self._tick_format
+
+    @tick_format.setter
+    def tick_format(self, format):
+        self._tick_format = str(format)
+
+    @property
+    def tick_side(self):
+        """Whether the ticks are on the left or right of the line.
```

Comment:
what about top and bottom?

Reply:
See the line below: "Imagine standing on the start position, with the line in front of you." The actual positioning depends on the angle, and is such that there are no jumps "over" the line. 

It's a bit like "starboad" and "port side" on ships 😄

Reply:
But I understand the confusion. I wonder if we can come up with a better term.